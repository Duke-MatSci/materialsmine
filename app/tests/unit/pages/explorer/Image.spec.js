import createWrapper from '../../../jest/script/wrapper'
import { enableAutoDestroy, RouterLinkStub } from '@vue/test-utils'
import ImageGallery from '@/pages/explorer/image/Image.vue'

const apollo = {
  images: {
    totalItems: 1,
    pageSize: 1,
    pageNumber: 1,
    totalPages: 1,
    hasPreviousPage: false,
    hasNextPage: true,
    images: [
      {
        file: '/api/files/583e05cfe74a1d205f4e2177',
        description: 'height image',
        microscopyType: 'AFM',
        type: 'grayscale',
        dimension: {
          width: 20,
          height: 20
        },
        metaData: {
          title: 'Morphology and electrical resistivity of melt mixed blends of polyethylene and carbon nanotube filled polycarbonate',
          doi: '10.1016/j.polymer.2003.10.003',
          id: '583e0672e74a1d205f4e21be',
          keywords: [
            'Blends',
            'Multiwalled carbon nanotube composites',
            'Electrical resistivity'
          ]
        }
      }
    ]
  }
}
describe('Image.vue', () => {
  let wrapper
  beforeEach(async () => {
    wrapper = await createWrapper(ImageGallery, {
      mocks: {
        $apollo: {
          loading: false,
          queries: {
            images: {
              refetch: jest.fn()
            },
            searchImages: {
              refetch: jest.fn()
            }
          }
        }
      }
    }, false)
    await wrapper.setData({ images: apollo.images })
  })

  enableAutoDestroy(afterEach)

  // TODO: TEST
  it('loads page', async () => {
    expect.assertions(1)
    expect(await wrapper.find('.gallery').exists()).toBe(true)
  })

  it('shows number of results', async () => {
    expect.assertions(2)
    expect(await wrapper.find('.utility-roverflow').exists()).toBe(true)
    expect(await wrapper.find('#css-adjust-navfont > span').text()).toMatch(
      /[1-9]\d*|No result/
    )
  })

  it('provides links for each result', async () => {
    const images = wrapper.vm.images.images
    expect.assertions(images.length)
    for (const [idx, image] of images.entries()) {
      expect(
        await wrapper.findAllComponents(RouterLinkStub).at(idx).props().to
      ).toEqual({
        name: 'ImageDetailView',
        params: { id: image.metaData.id, fileId: image.file }
      })
    }
  })

  it('limits result to page size', async () => {
    expect.assertions(1)
    await wrapper.setData({ pageSize: 5 })
    expect(wrapper.vm.images.images.length).toBeLessThanOrEqual(wrapper.vm.pageSize)
  })

  it('renders image title and description', () => {
    expect.assertions(6)
    const image = wrapper.find('.gallery-item')
    expect(image.exists()).toBe(true)
    expect(image.html()).toContain('Morphology and electrical')
    expect(image.find('.md-subheading').exists()).toBe(true)
    expect(image.find('.md-subheading').text()).not.toBe('')
    expect(image.find('.md-body-1').exists()).toBe(true)
    expect(image.find('.md-body-1').text()).not.toBe('')
  })

  it('mounts pagination component with the right parameters and calls the right method', async () => {
    expect.assertions(3)
    var spy = jest.spyOn(wrapper.vm, 'loadPrevNextImage').mockImplementation(() => {})
    const pagination = wrapper.findComponent('pagination-stub')
    expect(pagination.exists()).toBe(true)
    expect(pagination.props().cpage).toBe(wrapper.vm.pageNumber)
    await pagination.vm.$emit('go-to-page', 1)
    expect(spy).toHaveBeenCalledWith(1)
  })

  it('renders searchbox', () => {
    expect(wrapper.find('.search_box').exists()).toBe(true)
    expect(wrapper.find('.search_box').html()).toContain('filter')
    const searchBtn = wrapper.findAll('button').at(0)
    expect(searchBtn.text()).toBe('Search Images')
  })

  it('renders the right text when response empty', async () => {
    var expectedString = 'Sorry! No Image Found'
    const emptyData = apollo.images
    emptyData.images = []
    emptyData.totalItems = 0
    await wrapper.setData({ images: emptyData })
    expect(wrapper.find('.gallery-item').exists()).toBe(false)
    expect(wrapper.find('.utility-roverflow.u_centralize_text.u_margin-top-med').exists()).toBe(true)
    expect(wrapper.find('h1.visualize_header-h1.u_margin-top-med').text()).toBe(expectedString)
  })
})
