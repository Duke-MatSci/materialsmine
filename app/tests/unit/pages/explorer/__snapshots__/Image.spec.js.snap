// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image.vue renders pagination 1`] = `
<div class="gallery">
  <div class="utility-roverflow u--margin-toplg">
    <div class="search_box card-icon-container u--margin-toplg">
      <form class="form">
        <div class="search_box_form">
          <div class="form__group search_box_form-item-1"><input type="text" placeholder="Search Microstructure Image" name="search" id="search" required="required" class="form__input form__input--flat"> <label htmlfor="search" class="form__label search_box_form_label">Search Microstructure Image</label></div>
        </div>
        <div class="form__group search_box_form-item-2 explorer_page-nav u--margin-neg">
          <div class="form__field md-field"><select name="filter" id="filter" class="form__select">
              <option value="" disabled="disabled" selected="selected" hidden="hidden">Filter by...</option>
              <option value="Keyword">Keyword</option>
              <option value="filterByFiller">Filler</option>
              <option value="filterByYear">Year</option>
              <option value="filterByDOI">DOI</option>
              <option value="filterByMicroscopy">Microscopy</option>
            </select></div> <button type="submit" class="btn btn--primary btn--noradius search_box_form_btn mid-first-li display-text u--margin-pos">
            Search Images
          </button>
          <!---->
        </div>
      </form>
    </div>
  </div>
  <div class="utility-roverflow">
    <div class="u_content__result u_margin-top-small"><span id="css-adjust-navfont" class="u_color utility-navfont"><strong>Showing all images</strong> <span>
            1 result
          </span> <span class="utility-absolute-input"><label for="pagesize"><strong>Page size:</strong></label> <input placeholder="Enter page size" type="number" id="pagesize" name="pagesize" title="specify number of items per size" min="1" max="20" class="u_width--xs utility-navfont"></span></span></div>
    <div class="gallery-grid grid grid_col-5">
      <div class="md-card btn--animated gallery-item md-theme-default"><a>
          <div class="md-card-media-cover md-solid">
            <div class="md-card-media md-ratio-4-3"><img src="http://localhost/api/files/583e05cfe74a1d205f4e2177" alt="Morphology and electrical resistivity of melt mixed blends of polyethylene and carbon nanotube filled polycarbonate" crossorigin="Anonymous"></div>
            <div class="md-card-area u_gridbg">
              <div class="md-card-header u_show_hide"><span class="md-subheading"><strong> height image...</strong></span> <span class="md-body-1"> Morphology and electrical resistivity of melt mixed blends of polyethylene and carbon nanotube filled polycarbonate...</span></div>
            </div>
            <!---->
          </div>
        </a></div>
    </div>
    <div class="viz-pagination-container"><span class="viz-pagination u_centralize_content"><div class="pagination-label">
       Page
    </div> <div class="md-field pagination-field u_width--xs md-theme-default md-required md-has-value"><input type="text" id="pageInput" required="required" class="md-input"> <span class="md-error">Out of range</span>
      <!---->
      <transition-stub name="md-input-action" appear="true">
        <!---->
      </transition-stub>
      <transition-stub name="md-input-action" appear="true">
        <!---->
      </transition-stub>
    </div>
    <div class="pagination-label">
      of 1
    </div></span>
    <div class="viz-pagination viz-pagination_grid-4"><button disabled="disabled" class="pagination-button pagination-button-home">Home</button> <button disabled="disabled" class="pagination-button pagination-button-prev">Prev</button> <button disabled="disabled" class="pagination-button pagination-button-next">Next</button> <button disabled="disabled" class="pagination-button pagination-button-end">End</button></div>
  </div>
</div>
</div>
`;

exports[`Image.vue renders pagination 2`] = `
<div class="gallery">
  <div class="utility-roverflow u--margin-toplg">
    <div class="search_box card-icon-container u--margin-toplg">
      <form class="form">
        <div class="search_box_form">
          <div class="form__group search_box_form-item-1"><input type="text" placeholder="Search Microstructure Image" name="search" id="search" required="required" class="form__input form__input--flat"> <label htmlfor="search" class="form__label search_box_form_label">Search Microstructure Image</label></div>
        </div>
        <div class="form__group search_box_form-item-2 explorer_page-nav u--margin-neg">
          <div class="form__field md-field"><select name="filter" id="filter" class="form__select">
              <option value="" disabled="disabled" selected="selected" hidden="hidden">Filter by...</option>
              <option value="Keyword">Keyword</option>
              <option value="filterByFiller">Filler</option>
              <option value="filterByYear">Year</option>
              <option value="filterByDOI">DOI</option>
              <option value="filterByMicroscopy">Microscopy</option>
            </select></div> <button type="submit" class="btn btn--primary btn--noradius search_box_form_btn mid-first-li display-text u--margin-pos">
            Search Images
          </button>
          <!---->
        </div>
      </form>
    </div>
  </div>
  <div class="utility-roverflow">
    <div class="u_content__result u_margin-top-small"><span id="css-adjust-navfont" class="u_color utility-navfont"><strong>Showing all images</strong> <span>
            1 result
          </span> <span class="utility-absolute-input"><label for="pagesize"><strong>Page size:</strong></label> <input placeholder="Enter page size" type="number" id="pagesize" name="pagesize" title="specify number of items per size" min="1" max="20" class="u_width--xs utility-navfont"></span></span></div>
    <div class="gallery-grid grid grid_col-5">
      <div class="md-card btn--animated gallery-item md-theme-default"><a>
          <div class="md-card-media-cover md-solid">
            <div class="md-card-media md-ratio-4-3"><img src="http://localhost/api/files/583e05cfe74a1d205f4e2177" alt="Morphology and electrical resistivity of melt mixed blends of polyethylene and carbon nanotube filled polycarbonate" crossorigin="Anonymous"></div>
            <div class="md-card-area u_gridbg">
              <div class="md-card-header u_show_hide"><span class="md-subheading"><strong> height image...</strong></span> <span class="md-body-1"> Morphology and electrical resistivity of melt mixed blends of polyethylene and carbon nanotube filled polycarbonate...</span></div>
            </div>
            <!---->
          </div>
        </a></div>
    </div>
    <div class="viz-pagination-container"><span class="viz-pagination u_centralize_content"><div class="pagination-label">
       Page
    </div> <div class="md-field pagination-field u_width--xs md-theme-default md-required md-has-value"><input type="text" id="pageInput" required="required" class="md-input"> <span class="md-error">Out of range</span>
      <!---->
      <transition-stub name="md-input-action" appear="true">
        <!---->
      </transition-stub>
      <transition-stub name="md-input-action" appear="true">
        <!---->
      </transition-stub>
    </div>
    <div class="pagination-label">
      of 1
    </div></span>
    <div class="viz-pagination viz-pagination_grid-4"><button disabled="disabled" class="pagination-button pagination-button-home">Home</button> <button disabled="disabled" class="pagination-button pagination-button-prev">Prev</button> <button disabled="disabled" class="pagination-button pagination-button-next">Next</button> <button disabled="disabled" class="pagination-button pagination-button-end">End</button></div>
  </div>
</div>
</div>
`;
